[FLINK-20354] Rework standalone docs pages (#14346)
apache/flink: docs/deployment/resource-providers/standalone/index.md

under the License.
-->

> This page provides instructions on how to run Flink in a *fully distributed fashion* on a *static* (but possibly heterogeneous) cluster.

Reviewer: I don't like the "Overview" entry in the menu panel on the left as we're not giving an overview but introduce the basic (or native) deployment of a Standalone cluster. What is your opinion on the naming?

Author: I had exactly the same concern, however I didn't manage to rename it, without the navigation falling apart.

In the additional subpages of the standalone mode resource provider, we describe additional deployment methods which are based on the standalone mode: [Deployment in Docker containers]({% link deployment/resource-providers/standalone/docker.md %}), and on [Kubernetes]({% link deployment/resource-providers/standalone/kubernetes.md %}).

### Preparation

Flink runs on all *UNIX-like environments*, e.g. **Linux**, **Mac OS X**, and **Cygwin** (for Windows). Before you start to setup the system, make sure you have the fulfilled the following requirements.

- **Java 1.8.x** or higher installed,
- Downloaded a recent Flink distribution from the [download page]({{ site.download_url }}) and unpacked it.

### Starting a Standalone Cluster (Session Mode)

These steps show how to launch a Flink standalone cluster, and submit an example job:

{% highlight bash %}
# we assume to be in the root directory of the unzipped Flink distribution

# (1) Start Cluster
./bin/start-cluster.sh

# (2) You can now access the Flink Web Interface on http://localhost:8081

# (3) Submit example job

> ./bin/flink run ./examples/streaming/TopSpeedWindowing.jar

Author: but isn't it the purpose of the CLI documentation page to show how to pass job-related parameters?
This guide focuses on educating users on the standalone deployment.

# (4) Stop the cluster again
./bin/stop-cluster.sh
{% endhighlight %}

In step `(1)`, we've started 2 processes: A JVM for the JobManager, and a JVM for the TaskManager. The JobManager is serving the web interface accessible at [localhost:8081](http://localhost:8081).
In step `(3)`, we are starting a Flink Client (a short-lived JVM process) that submits an application to the JobManager.

> ## Deployment Modes Supported by the Standalone Cluster

Reviewer: This also includes updating the headline as suggested above.

> ## Deployment Modes Supported by the Standalone Cluster

Author: Damn, forgot that one.

### Configuration

All available configuration options are listed on the [configuration page]({% link deployment/config.md %}), in particular the [Basic Setup]({% link deployment/config.md %}#basic-setup) section contains good advise on configuring the ports, memory, parallelism etc.

### Debugging

If Flink is behaving unexpectedly, we recommend looking at Flink's log files as a starting point for further investigations.

The log files are located in the `logs/` directory. There's a `.log` file for each Flink service running on this machine.

Alternatively, logs are available from the Flink web frontend (both for the JobManager and each TaskManager).

By default, Flink is logging on the "INFO" log level, which provides basic information for all obvious issues. For cases where Flink supposedly behaving wrongly, reducing the log level to "DEBUG" is advised. The logging level is controlled via the `conf/log4.properties` file.
Setting `rootLogger.level = DEBUG` will boostrap Flink on the DEBUG log level. Note that a restart of Flink is required for the changes to take effect.

There's a dedicated page on the [logging]({%link deployment/advanced/logging.md %}) in Flink.

> ### The start and stop scripts

Author: Good idea! Applied.

#### start-cluster.sh

The scripts provided with the standalone mode (in the `bin/` directory) use the `conf/workers` and `conf/masters` files, to determine the number of cluster instances to start and stop with the `bin/start-cluster.sh` and `bin/stop-cluster.sh` scripts.

If password-less ssh access to the listed machines is configured, and they share the same directory structure, the script also supports starting and stopping instances remotely.

**Example 1: Start a cluster with 2 TaskManagers locally**

`conf/masters` contents:
{% highlight bash %}
localhost
{% endhighlight %}

`conf/workers` contents:
{% highlight bash %}
localhost
localhost
{% endhighlight %}

> **Example 2: Start a distributed cluster JobMangers**

Author: I would consider setting up a distributed standalone cluster a fairly advanced topic. Most of our users are probably using K8s or YARN for deployment.
I'm concerned that this will confuse new users.

{% highlight bash %}
worker1
worker2
worker3
{% endhighlight %}

Note that the configuration key `jobmanager.rpc.address` needs to be set to `master1` for this to work.

We show a third example with a standby JobManager in the [high-availability section](#setting-up-high-availability).

#### (jobmanager|taskmanager).sh

The `bin/jobmanager.sh` and `bin/taskmanager.sh` script support starting the respective daemon in the background (using the `start` argument), or in the foreground (using `start-foreground`). In the foreground mode, the logs are printed to standard out. This mode is useful for deployment scenarios where another process is controlling the Flink daemon (e.g. Docker).

The scripts can be called multiple times, for example if multiple TaskManagers are needed. The instances are tracked by the scripts, and can be stopped one-by-one (using `stop`) or all together (using `stop-all`).

#### Windows Cygwin Users

If you are installing Flink from the git repository and you are using the Windows git shell, Cygwin can produce a failure similar to this one:

{% highlight bash %}
c:/flink/bin/start-cluster.sh: line 30: $'\r': command not found
{% endhighlight %}

> This error occurs because git is automatically transforming UNIX line endings to Windows style line endings when running in Windows. The problem is that Cygwin can only deal with UNIX style line endings. The solution is to adjust the Cygwin settings to deal with the correct line endings by following these three steps:

Author: I think you are right. Stuff runs on windows 😕

1. Start a Cygwin shell.

2. Determine your home directory by entering

{% highlight bash %}
cd; pwd
{% endhighlight %}

    This will return a path under the Cygwin root path.

3. Using NotePad, WordPad or a different text editor open the file `.bash_profile` in the home directory and append the following: (If the file does not exist you will have to create it)

{% highlight bash %}
export SHELLOPTS
set -o igncr
{% endhighlight %}

> Save the file and open a new bash shell.

Author: I don't have Windows available either. I'll add the 4. but leave the sourcing out.

under the License.
-->

* This will be replaced by the TOC
{:toc}

## Getting Started

> This *Getting Started* section guides you through the local setup (on one machine, but in separate processes) of a Flink cluster. This can easily be expanded to set up a distributed standalone cluster, which we describe in the [reference section](#the-start-and-stop-scripts).

Reviewer: ```suggestion
This *Getting Started* section guides you through the local setup (on one machine, but in separate processes) of a Flink cluster. This can easily be expanded to set up a distributed standalone cluster, which we describe in the [reference section](#example-2-start-a-distributed-cluster-jobmangers).
```

We demonstrate this mode by locally starting the `TopSpeedWindowing.jar` example, running on a single TaskManager.

The application jar file needs to be available in the classpath. The easiest approach to achieve that is putting the jar into the `lib/` folder:

{% highlight bash %}
$ cp ./examples/streaming/TopSpeedWindowing.jar lib/
{% endhighlight %}

Then, we can launch the JobManager:

{% highlight bash %}
$ ./bin/standalone-job.sh start --job-classname org.apache.flink.streaming.examples.windowing.TopSpeedWindowing
{% endhighlight %}

The web interface is now available at [localhost:8081](http://localhost:8081). However, the application won't be able to start, because there are no TaskManagers running yet:

{% highlight bash %}
$ ./bin/taskmanager.sh start
{% endhighlight %}

Note: You can start multiple TaskManagers, if your application needs more resources.

Stopping the services is also supported via the scripts. Call them multiple times if you want to stop multiple instances, or use `stop-all`:

{% highlight bash %}
$ ./bin/taskmanager.sh stop
$ ./bin/standalone-job.sh stop
{% endhighlight %}

### Per-Job Mode

> Per-Job Mode is not supported by the the Standalone Cluster.

Reviewer: ```suggestion
Per-Job Mode is not supported by the Standalone Cluster.
```

1. Start a Cygwin shell.

2. Determine your home directory by entering

    ```bash
    cd; pwd
    ```

    This will return a path under the Cygwin root path.

3. Using NotePad, WordPad or a different text editor open the file `.bash_profile` in the home directory and append the following (if the file does not exist you will have to create it):

    ```bash
    $ export SHELLOPTS
    $ set -o igncr
    ```

4. Save the file and open a new bash shell.

### Setting up High-Availability

In order to enable HA for a standalone cluster, you have to use the [ZooKeeper HA services]({% link deployment/ha/zookeeper_ha.md %}).

Additionally, you have to configure your cluster to start multiple JobManagers.

In order to start an HA-cluster configure the *masters* file in `conf/masters`:

- **masters file**: The *masters file* contains all hosts, on which JobManagers are started, and the ports to which the web user interface binds.

  <pre>
master1:webUIPort1
[...]
masterX:webUIPortX
  </pre>

> By default, the job manager will pick a *random port* for inter process communication. You can change this via the [high-availability.jobmanager.port]({% link deployment/config.md %}#high-availability-jobmanager-port) key. This key accepts single ports (e.g. `50010`), ranges (`50000-50025`), or a combination of both (`50010,50011,50020-50025,50050-50075`).

Reviewer: ```suggestion
By default, the JobManager will pick a *random port* for inter process communication. You can change this via the [high-availability.jobmanager.port]({% link deployment/config.md %}#high-availability-jobmanager-port) key. This key accepts single ports (e.g. `50010`), ranges (`50000-50025`), or a combination of both (`50010,50011,50020-50025,50050-50075`).
```

> #### Example: Standalone HA Cluster with 2 JobManagers

Reviewer: Can we switch from `<pre/>` tags to "\`\`\`sh \`\`\`"? This way, we would enable syntax highlighting for these code blocks.

> 1. **Configure high availability mode and ZooKeeper quorum** in `conf/flink-conf.yaml`:

Reviewer: I think we could remove the formatting for the items.
